# Big Data: Tools and Methods

# Assignment 1 - Xin Zhou

#1. If we list all the natural numbers below 20 that are multiples of 3 or 5 but not multiples both of 3 and 5, we get 3, 5, 6, 9, 10, 12, 18. The sum of these multiples is 63. Write a function, sum_multiples_3_5, that returns the sum of the multiples of 3 and 5 less than N.

In [2]:
def sum_multiples_3_5(n:Int ): Int ={
  var sum:Int = 0
  for(i <- 0 to n-1){
    if(i%3 == 0 && i%5 != 0){sum += i }
    if(i%5 == 0 && i%3 != 0){sum += i }

  }
  return sum
}
sum_multiples_3_5(20)

defined function sum_multiples_3_5
res1_1: Int = 63

#2. Write a function, patternCount with two arguments. The first arguments is a string, lets call it text, and the second argument is also a string, call it pattern. The function patternCount returns the number of times the pattern occurs in the text. For example
pattern_count(“abababa”, “aba”) == 3 
pattern_count(“aaaaa”, “aa”) == 4 
pattern_count(“Abcde”, “abc”) == 0

In [3]:
def pattern_count(text:String, pattern:String): Int={
    var tLen = text.length()
    var pLen = pattern.length()
    var count = 0;
    for(index <- 0 to tLen-pLen){
        if(text.substring(index, index+pLen)==pattern){count += 1 }
    }
    return count
}
pattern_count("abababa", "aba")
pattern_count("aaaaa", "aa")
pattern_count("Abcde", "abc")

defined function pattern_count
res2_1: Int = 3
res2_2: Int = 4
res2_3: Int = 0

#3. Write a function product(s:String) that returns the product of the Unicode codes of all letters in string s without using a loop.

In [15]:
def product(s: String): String = { 
    var arr = s.map(_.toInt)map(_.toString)
    var p = BigInt("1")
    for (n <- arr){
        p *= BigInt(n)
    }
    return p.toString;
}
product("Bigdata")

defined function product
res14_1: String = "77906181276000"

#4. Write a function no Duplicates (ints:Array[Int]) that returns an array all values from the array ints with duplicates removed.

In [18]:
def noDuplicates(ints:Array[Int]): Array[Int] = {
    return ints.toSet.toArray
}
var result = noDuplicates(Array(1,2,2,3,4,4,5))

defined function noDuplicates
result: Array[Int] = Array(5, 1, 2, 3, 4)

#5. Create an array that contains the value 0.00001f repeated 1000000 times. Sum up all the elements of the array. What is the result? How big is the error?

In [33]:
var arr = Array.fill(1000000){0.00001f}
var arrSum = arr.sum
var error = BigDecimal(10.0f) - BigDecimal(arrSum)

arr: Array[Float] = Array(
  1.0E-5F,
  1.0E-5F,
  1.0E-5F,
  1.0E-5F,
  1.0E-5F,
  1.0E-5F,
  1.0E-5F,
  1.0E-5F,
  1.0E-5F,
  1.0E-5F,
  1.0E-5F,
...
arrSum: Float = 9.917345F
error: BigDecimal = 0.08265495300293

#6. Using Scala compute the sums 0.0001f + 8000.0f and 0.000000001 + 90000000. What is the result? How big is the error?

In [20]:
var r1 = 0.0001f + 8000.0f 
println("sums 0.0001f + 8000.0f = " + r1.toString+"f")
var err1 = BigDecimal(0.0001f)+BigDecimal(8000.0f) - BigDecimal(r1)
println("Error for 0.0001 + 8000.0 is "+err1.toString)

var r2 = 0.000000001 + 90000000
println("sums 0.000000001 + 90000000 = " + r2.toString)
var err2 = BigDecimal(0.000000001)+BigDecimal(90000000) - BigDecimal(r2)
println("Error for 0.000000001 + 90000000 is "+err2.toString)

sums 0.0001f + 8000.0f = 8000.0f
Error for 0.0001 + 8000.0 is 0.00009999999747378752
sums 0.000000001 + 90000000 = 9.0E7
Error for 0.000000001 + 90000000 is 1.0E-9


r1: Float = 8000.0F
err1: BigDecimal = 0.00009999999747378752
r2: Double = 9.0E7
err2: BigDecimal = 1.0E-9

#7. Write a function random Ints(n:Int,range:Int) that returns an array of n integers all between 0 and range.

In [9]:
def randomInts(n:Int,range:Int): Array[Int]={
    var result = Array.fill(n){0}
    val r = scala.util.Random
    for(i <- 0 to n-1){
        result(i) = r.nextInt(range)
    }
    return result
}
var result = randomInts(5, 50)

defined function randomInts
result: Array[Int] = Array(10, 14, 39, 48, 5)

#8. Create three files with random integers in the range 0 to Int.MaxValue/2. The first file contains 1,000 integers, the second 100,000 integers, the third 10,000,000 integers. Write scala code to find the average and median value for each file. What are the average and median value for each file? How long does it take compute those values for each file?

In [32]:
import java.io._
val pw1 = new PrintWriter(new File("file1.txt" ))
var result1 = randomInts(1000, Int.MaxValue/2)
pw1.write(result1.mkString(" "))
pw1.close
val pw2 = new PrintWriter(new File("file2.txt" ))
var result2 = randomInts(100000, Int.MaxValue/2)
pw2.write(result2.mkString(" "))
pw2.close
val pw3 = new PrintWriter(new File("file3.txt" ))
var result3 = randomInts(10000000, Int.MaxValue/2)
pw3.write(result3.mkString(" "))
pw3.close
import scala.io.Source._
val lines1 = fromFile("file1.txt").getLines
val lines2 = fromFile("file2.txt").getLines
val lines3 = fromFile("file3.txt").getLines

def rollingAvgMedian(filename:String){
    val lines = fromFile(filename).getLines
    for (line <- lines){
        val t0 = System.currentTimeMillis()
        var arr: Array[Int] = line.split(" ").map(_.toInt)
        arr.sortWith(_ < _)
        var size = arr.length
        var N = size.toDouble
        var avg:Double = 0
        for (number <- arr){
            avg += number.toDouble/N
        }
        var median = (BigInt(arr(size/2-1))+ BigInt(arr(size/2)))/2  //find the middle two numbers in sorted array, in order to calculate median     
        val t1 = System.currentTimeMillis()
        println("For "+filename+" :")
        println("Average is "+avg)
        println("Median is "+median)
        if (t1 - t0 < 1000){
            println("Elapsed time: " + (t1 - t0) + "ms")
        }
        else{println("Elapsed time: " + (t1 - t0)/1000 + "s")}      
    }
    println("-------------------------------")
}
rollingAvgMedian("file1.txt")
rollingAvgMedian("file2.txt")
rollingAvgMedian("file3.txt")

For file1.txt :
Average is 5.3156646540999985E8
Median is 324975516
Elapsed time: 3ms
-------------------------------
For file2.txt :
Average is 5.373574214714749E8
Median is 459393951
Elapsed time: 66ms
-------------------------------
For file3.txt :
Average is 5.368236262546402E8
Median is 116572276
Elapsed time: 13s
-------------------------------


import java.io._

pw1: PrintWriter = java.io.PrintWriter@14f3f193
result1: Array[Int] = Array(
  545192727,
  624557298,
  95299628,
  122266929,
  816650112,
  1068926765,
  373675137,
  567153187,
  44229320,
  917415645,
  919919289,
...
pw2: PrintWriter = java.io.PrintWriter@73707b48
result2: Array[Int] = Array(
  543589590,
  705819311,
  1010871186,
  951255496,
  815447904,
  827608908,
  987117944,
  137391314,
  382123445,
  747057312,
  1033397032,
...
pw3: PrintWriter = java.io.PrintWriter@13a498a
result3: Array[Int] = Array(
  914543214,
  649428138,
  1009523164,
  815575542,
  252029548,
  598349660,
  625123325,
  62963660,
  1073257913,
  1002904058,
  590008908,
...
import scala.io.Source._

lines1: Iterator[String] = non-empty iterator
lines2: Iterator[String] = non-empty iterator
lines3: Iterator[String] = non-empty iterator
defined function rollingAvgMedian

#8 Results & Comments
--------------------------------
Using intgers: such as array of 1,2,3,....10, average = 1/10 + 2/10 + ...10/10 accumulating<br>
For file1.txt :<br>
Average is 5.3156646540999985E8<br>
Median is 324975516<br>
Elapsed time: 3ms<br>
<br>
For file2.txt :<br>
Average is 5.373574214714749E8<br>
Median is 459393951<br>
Elapsed time: 66ms<br>
<br>
For file3.txt :<br>
Average is 5.368236262546402E8<br>
Median is 116572276<br>
Elapsed time: 13s<br>
<br>
Using BigDecimal would result in 28 seconds for dealing with file 3


#9. Write a function min max (values:Array[Int]) that returns a pair containing the smallest and the largest values in the array.

In [73]:
def minmax(values:Array[Int]): Array[Int]={
    if(values.isEmpty) throw new java.lang.UnsupportedOperationException("array is empty")
    var pair = Array(Int.MaxValue, Int.MinValue)
    for(n <- values){
        pair(0) = math.min(n, pair(0))
        pair(1) = math.max(n, pair(1))
    }
    return pair
}
var p = minmax(Array(2,1,5,8,9,3,4,10,6))

defined function minmax
p: Array[Int] = Array(1, 10)

#10. Make a class Car with read-only properties for manufacturer, model name, and model year, and a read-write property for the license plate. Supply four constructors. All require the manufacturer and model name. Optionally, model year and license plate can also be specified in the constructor. If not, the model year is set to -1 and the license plate to the empty string. Which constructor are you choosing as the primary constructor? Why?

In [21]:
class Car(manufacturer: String, modelName: String, modelYear: Int = -1, licensePlace: String = "") {
    private var _manufacturer = manufacturer
    private var _modelName = modelName
    private var _modelYear = modelYear
    var license = licensePlace
}
var myCar = new Car("Toyota", "Camry")
var myCar2 = new Car("Toyota", "Camry", 2014, "1348485")

defined class Car
myCar: Car = $sess.cmd20Wrapper$Helper$Car@2da2b241
myCar2: Car = $sess.cmd20Wrapper$Helper$Car@3b05f3dd

Comment:<br>
My primary construtor is the class body, because it is more convinient than the structure in Java or C++, it can pass down variables and store in the created class directly. 